sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp

```diff
@@ -0,0 +1,367 @@
+template <typename Group, typename T, class BinaryOperation>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<ext::oneapi::experimental::is_fixed_size_group_v<Group> &&
+                         std::is_integral_v<T>,
+                     T>
+    masked_reduction_cuda_shfls(Group g, T x, BinaryOperation binary_op,
+                                const uint32_t MemberMask) {
+
+  for (int i = g.get_local_range()[0] / 2; i > 0; i /= 2) {
+    auto tmp = __nvvm_shfl_sync_bfly_i32(MemberMask, x, -1, i);
+    x = binary_op(x, tmp);
+  }
+  return x;
+}
+
+template <typename Group, typename T>
+inline __SYCL_ALWAYS_INLINE std::enable_if_t<
+    ext::oneapi::experimental::is_user_constructed_group_v<Group>, T>
+non_uniform_shfl_T(const uint32_t MemberMask, T x, int shfl_param) {
+  if constexpr (ext::oneapi::experimental::is_fixed_size_group_v<Group>) {
+    return __nvvm_shfl_sync_up_i32(MemberMask, x, shfl_param, 0);
+  } else {
+    return __nvvm_shfl_sync_idx_i32(MemberMask, x, shfl_param, 31);
+  }
+}
+
+template <typename Group, typename T>
+inline __SYCL_ALWAYS_INLINE std::enable_if_t<
+    ext::oneapi::experimental::is_user_constructed_group_v<Group>, T>
+non_uniform_shfl(Group g, const uint32_t MemberMask, T x, int shfl_param) {
+  T res;
+  if constexpr (std::is_same_v<T, double>) {
+    int x_a, x_b;
+    asm volatile("mov.b64 {%0,%1},%2; \n\t" : "=r"(x_a), "=r"(x_b) : "l"(x));
+
+    auto tmp_a = non_uniform_shfl_T<Group>(MemberMask, x_a, shfl_param);
+    auto tmp_b = non_uniform_shfl_T<Group>(MemberMask, x_b, shfl_param);
+    asm volatile("mov.b64 %0,{%1,%2}; \n\t"
+                 : "=l"(res)
+                 : "r"(tmp_a), "r"(tmp_b));
+  } else {
+    auto input = std::is_same_v<T, float> ? __nvvm_bitcast_f2i(x) : x;
+    auto tmp_b32 = non_uniform_shfl_T<Group>(MemberMask, input, shfl_param);
+    res = std::is_same_v<T, float> ? __nvvm_bitcast_i2f(tmp_b32) : tmp_b32;
+  }
+  return res;
+}
+
+// Opportunistic/Ballot group reduction using shfls
+template <typename Group, typename T, class BinaryOperation>
+inline __SYCL_ALWAYS_INLINE std::enable_if_t<
+    ext::oneapi::experimental::is_user_constructed_group_v<Group> &&
+        !ext::oneapi::experimental::is_fixed_size_group_v<Group>,
+    T>
+masked_reduction_cuda_shfls(Group g, T x, BinaryOperation binary_op,
+                            const uint32_t MemberMask) {
+  if (MemberMask == 0xffffffff) {
+    for (int i = 16; i > 0; i /= 2) {
+      auto tmp = __nvvm_shfl_sync_bfly_i32(MemberMask, x, -1, i);
+      x = binary_op(x, tmp);
+    }
+    return x;
+  }
+
+  unsigned localSetBit = g.get_local_id()[0] + 1;
+
+  // number of elements requiring binary operations each loop iteration
+  auto opRange = g.get_local_range()[0];
+
+  // remainder that won't have a binary partner each loop iteration
+  int remainder;
+
+  while (opRange / 2 >= 1) {
+    remainder = opRange % 2;
+
+    // stride between local_ids forming a binary op
+    int stride = opRange / 2;
```

Comment:
Nit: Any reason for this (and `unfoldedSrcSetBit`) to be signed?

Reply:
No they can be unsigned. I will change this.

Reply:
Note however that `unfoldedSrcSetBit` has to be signed in `masked_scan_cuda_shfls`

Reply:
If it makes sense to keep it signed, it's no big deal. 😄

Reply:
OK I've kept just `unfoldedSrcSetBit` signed, consistent with the function where this variable name needs to be signed.